### PR TITLE
redux + formik + yup intregation

### PR DESCRIPTION
**Pull request by sujaythakurata**
### **redux + formik + yup integration for form 1**
create individual input, radio, select tag and also localstorage save data & get data if any on load
image
image
image

### CODE REVIEW

File: hooks/useForm5.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import { IForm5 } from '../types/form';
+
+export default function useForm5() {
+  const formData = useAppSelector(state => state.form.formData.form5);
+  const curForm = useAppSelector(state => state.form.curForm);
+  const dispatch = useAppDispatch();
+  const formik: FormikProps<IForm5> = useFormik({
+    initialValues: formData,
+    validationSchema: Yup.object({ ...validation }),
+    onSubmit: values => {
```
Comment (mandeepduggal) [inline hooks/useForm5.ts]: Should not use inline functions

File: components/form.tsx [hunk 2]
```diff
@@ -1,7 +1,18 @@
...
+import Preview from './preview';
+
 export default function Form() {
-  return (
-    <>
-      <h1 data-testid="form">Form</h1>
-    </>
-  );
+  const curState = useAppSelector(state => state.form.curForm);
+  if (curState === env.form.form1.state) return <Form1 />;
```
Comment (mandeepduggal) [inline components/form.tsx]: use switch case if possible

File: components/fields/input.tsx [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import { FormikProps } from 'formik';
+import { ChangeEvent } from 'react';
+
+type IInput = FormikProps<any> & {
```
Comment (nakulnagariy) [inline components/fields/input.tsx:6]: Keep some sensable names, e.g. `InputComp`. this applies to every other file.

File: components/fields/nextbtn.tsx [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import style from '../../styles/home.module.scss';
+
+export default function NextBtn() {
```
Comment (nakulnagariy) [inline components/fields/nextbtn.tsx]: naming here again.

File: components/fields/radio.tsx [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import { FormikProps } from 'formik';
+import { ChangeEvent } from 'react';
+
+type IRadio = FormikProps<any> & {
+  id: string;
+  label: string;
+  // eslint-disable-next-line no-unused-vars
+  onBlur: (event: ChangeEvent<any>) => void;
```
Comment (nakulnagariy) [inline components/fields/radio.tsx:9]: try to use the right event to avoid `any` wherever possible.

File: components/fields/radio.tsx [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+export default function Radio(props: IRadio) {
+  return (
+    <>
+      <label className="m-1">{props.label}</label>
+      {props.formValue.map((value, index) => (
+        <div className="form-check form-check-inline" key={index}>
+          <input
+            className="form-check-input"
+            type="radio"
+            name={props.id}
```
Comment (nakulnagariy) [inline components/fields/radio.tsx]: can we keep the name other than id, would be more meaningful?

File: components/form.tsx [hunk 7]
```diff
@@ -1,7 +1,18 @@
+import env from '../env/env.json';
+import { useAppSelector } from '../states/hook';
+import Form1 from './form_1/form_1';
```
Comment (nakulnagariy) [inline components/form.tsx]: naming again, keep sectioned wise, ownerDetailsForm, TenantDetailsForm, otherDetails more descriptive.

File: components/form_1/form_1.tsx [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import { FormikProps } from 'formik';
+import { memo } from 'react';
+import useForm1 from '../../hooks/useForm1';
+import useInitialData from '../../hooks/useInitialdata';
+import style from '../../styles/home.module.scss';
+import { IForm1 } from '../../types/form';
+import NextBtn from '../fields/nextbtn';
+import Form1Fields from './form_1_fields';
+
+function Form1() {
```
Comment (nakulnagariy) [inline components/form_1/form_1.tsx]: use arrow functions for creating components, and export in the same line wherever possible.

File: components/form_1/form_1_fields.tsx [hunk 9]
```diff
@@ -0,0 +1,94 @@
+import { FormikProps } from 'formik';
+import { ChangeEvent } from 'react';
+import style from '../../styles/home.module.scss';
+import env from '../../env/env.json';
```
Comment (nakulnagariy) [inline components/form_1/form_1_fields.tsx]: env.json, name & use of this file is not right, env files usually used for keeping the secrets like KEYS, PASSWORD, DB details etc but not for config files, maybe you want to have something like constant.js or config.js

File: functions/form1validation.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
+import * as Yup from 'yup';
+import env from '../env/env.json';
+
+/*
+ *@return yup object validation schemea
+ *form 1 validation
+ */
+
+const [key1, key2, key3, key4, key5] = env.form.form1.keys;
```
Comment (nakulnagariy) [inline functions/form1validation.ts]: didn't know what key1,2,3 means, use the appropriate names.

File: functions/form1validation.ts [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+type form1 = Yup.InferType<any>;
+const validation: form1 = {};
+validation[key1] = Yup.string().required('this field is required');
+validation[key2] = Yup.string()
+  .matches(/[a-zA-z ]{1,}$/, 'valid name only letters allowed')
```
Comment (nakulnagariy) [inline functions/form1validation.ts]: you can have your regex and messages in the same constant file.

File: types/formikForm.ts [hunk 12]
```diff
@@ -0,0 +1 @@
+export default 'hello';
```
Comment (nakulnagariy) [inline types/formikForm.ts]: ????? for what?

**Review by nakulnagariy — CHANGES_REQUESTED** [review]
1. Progress bar is missing.
2. Naming of files, the variable is not right.
3. keep the constant and other strings in the constant file.
4. Progress bar is missing.
5. don't add the next, prev button in each section.
6. you can have multiple components for sections but display them on one screen, maybe in accordion.
7. test cases are missing.
8. again don't focus on creating a complete application in one go, have one section and complete that.

File: hooks/useForm5.ts (hunk 1 above)
Comment (sujaythakurata) [reply hooks/useForm5.ts]: changes done

File: components/form.tsx (hunk 2 above)
Comment (sujaythakurata) [reply components/form.tsx]: changes done

File: components/fields/radio.tsx (hunk 5 above)
Comment (sujaythakurata) [reply components/fields/radio.tsx:9]: ok sure. but here I am using formik that's why I keep it any to use the function for every field

File: components/fields/radio.tsx (hunk 6 above)
Comment (sujaythakurata) [reply components/fields/radio.tsx]: ok I will change that

File: components/form.tsx (hunk 7 above)
Comment (sujaythakurata) [reply components/form.tsx]: changes done

File: components/form_1/form_1.tsx (hunk 8 above)
Comment (sujaythakurata) [reply components/form_1/form_1.tsx]: ok sure will do the changes

File: components/form_1/form_1_fields.tsx (hunk 9 above)
Comment (sujaythakurata) [reply components/form_1/form_1_fields.tsx]: ok sure I will change the name

File: functions/form1validation.ts (hunk 10 above)
Comment (sujaythakurata) [reply functions/form1validation.ts]: changes done

File: functions/form1validation.ts (hunk 11 above)
Comment (sujaythakurata) [reply functions/form1validation.ts]: ok

File: types/formikForm.ts (hunk 12 above)
Comment (sujaythakurata) [reply types/formikForm.ts]: dummy file now it's changed.

File: components/fields/input.tsx (hunk 3 above)
Comment (sujaythakurata) [reply components/fields/input.tsx:6]: ok sure

Comment (sujaythakurata) [reply components/fields/input.tsx:6]: - progress bar is pending will be finish by tomorrow
- naming of file and variable changed
- env.json is the constant file I will change the file name to config.json
-  ok I will try to add is accordion way
- I will add testcases

**Pull request merged**